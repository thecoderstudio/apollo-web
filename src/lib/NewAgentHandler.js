export default class NewAgentHandler {
  constructor() {
    this.supportedArch = ['amd64', 'arm64', 'arm'];
    this.supportedOS = ['linux', 'darwin', 'freebsd', 'openbsd'];
  };

  getDirectlyOnMachineCommand(os, arch, agentId, secret, host) {
    let command = `curl http://localhost:1970/agent/download?target_os=${os}&target_arch=${arch}`;
    return command += ` ${this.getExecuteCommand(agentId, secret, host)}`;
  };

  getExecuteCommand(agentId, secret, host) {
    return `chmod +x apollo-agent.bin && ./apollo-agent.bin --agent-id=${agentId} --secret=${secret} --host=${host}`;
  };
};