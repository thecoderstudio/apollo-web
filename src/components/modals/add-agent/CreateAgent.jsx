import React from 'react';
import styled from 'styled-components';
import axios from 'axios';
import { Formik } from 'formik';
import { StatusCodes } from 'http-status-codes';
import PropTypes from 'prop-types';
import Dropdown from '../../Dropdown';
import LoadingButton from '../../buttons/LoadingButton';
import Input from '../../Input';
import OutlinedButton from '../../buttons/OutlinedButton';
import { createAgentSchema } from '../../../validation/agent';
import { parseHTTPErrors } from '../../../util/parser';
import { handleHTTPResponse } from '../../../actions/error';
import media from '../../../util/media';
import NewAgentHandler from '../../../lib/NewAgentHandler';

const propTypes = {
  createAgentSuccessCallback: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired
};

const TwoColumnGrid = styled.div`
  display: grid;
  grid-template-columns: 50% 50%;
  margin: 20px 0px 20px 0px;

  ${media.phone`
    grid-template-columns: [column-one] 100%;
    grid-template-rows: [row-one] 1fr [row-two] 1fr;
  `}
`;

const TextAndInputFieldWrapper = styled.div`
  display: grid;
  width: 100%;
  grid-template-columns: [text] 50% [dropdown] 50%;
  height: 100px;

  ${media.phone`
    height: 120px;
    grid-template-columns: [column-one] 100%;
    grid-template-rows: [text] 30px [dropdown] 85px;
  `}
`;

const InputFieldWrapper = styled.div`
  grid-column: dropdown;

  ${media.phone`
    grid-column: column-one;
    grid-row: dropdown;
    width: 100%;
  `}
`;

const StyledDropDown = styled(Dropdown)`
  width: 200px;
  float: right;

  ${media.phone`
    width: 100%;
  `}
`;

const TextWrapper = styled.p`
  grid-column: text;
  line-height: 75px;
  margin-top: 0px;
  margin-bottom: 0px;

  ${media.phone`
    width: 100%;
    line-height: 100%;
    grid-column: column-one;
    grid-row: text;
  `}
`;

const CreateAgentButton = styled(LoadingButton)`
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;

  ${media.phone`
    width: 100%;
    margin-top: 15px;
  `}
`;

const StyledInput = styled(Input)`
  font-family: 'B612', sans-serif;
  font-weight: 600;
  width: 200px;
  float: right;

  ${media.phone`
    width: 100%;
    grid-row: row-two;
  `}
`;

const CloseOutlinedButton = styled(OutlinedButton)`
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;

  ${media.phone`
    width: 100%;
    grid-row: row-two;
  `}
`;

class CreateAgent extends React.PureComponent {
  constructor(props) {
    super(props);
    this.newAgentHandler = new NewAgentHandler();
    this.state = {
      selectedOperatingSystem: this.newAgentHandler.supportedOS[0],
      selectedArchitecture: this.newAgentHandler.supportedArch[0]
    };
    this.createAgent = this.createAgent.bind(this);
    this.selectArchitecture = this.selectArchitecture.bind(this);
    this.selectOperatingSystem = this.selectArchitecture.bind(this);
  }

  createAgent(values, { setErrors }) {
    const { createAgentSuccessCallback } = this.props;
    const { selectedArchitecture, selectedOperatingSystem } = this.state;

    this.setState({ loading: true });
    axios.post(
      `${process.env.APOLLO_HTTP_URL}agent`,
      values,
      { withCredentials: true }
    )
      .then((response) => {
        createAgentSuccessCallback(
          response,
          selectedArchitecture,
          selectedOperatingSystem
        );
      })
      .catch((error) => {
        handleHTTPResponse(error.response, true, true);
        if (error.response.status === StatusCodes.BAD_REQUEST) {
          setErrors(parseHTTPErrors(error.response.data, { name: 'name' }));
        }
        this.setState({ loading: false });
      });
  }

  selectArchitecture(value) {
    this.setState({ selectedArchitecture: value });
  }

  selectOperatingSystem(value) {
    this.setState({ selectedOperatingSystem: value });
  }

  render() {
    const { selectedOperatingSystem, selectedArchitecture, loading } = this.state;
    const { onClose } = this.props;
    return (
      <Formik
        initialValues={{ name: '' }}
        validationSchema={createAgentSchema}
        onSubmit={this.createAgent}
      >
        {({
          values, errors, handleChange, handleSubmit
        }) => (
          <form onSubmit={handleSubmit}>
            <TextAndInputFieldWrapper>
              <TextWrapper>Agent name</TextWrapper>
              <InputFieldWrapper>
                <StyledInput
                  inverted
                  name="name"
                  type="name"
                  value={values.name}
                  error={errors.name}
                  placeholder="007"
                  onChange={handleChange}
                />
              </InputFieldWrapper>
            </TextAndInputFieldWrapper>
            <TextAndInputFieldWrapper>
              <TextWrapper>Operating system</TextWrapper>
              <InputFieldWrapper>
                <StyledDropDown
                  selected={selectedOperatingSystem}
                  options={this.newAgentHandler.supportedOS}
                  optionSelectedCallback={this.selectOperatingSystem}
                />
              </InputFieldWrapper>
            </TextAndInputFieldWrapper>
            <TextAndInputFieldWrapper>
              <TextWrapper>Architecture</TextWrapper>
              <InputFieldWrapper>
                <StyledDropDown
                  selected={selectedArchitecture}
                  options={this.newAgentHandler.supportedArch}
                  optionSelectedCallback={this.selectArchitecture}
                />
              </InputFieldWrapper>
            </TextAndInputFieldWrapper>
            <TwoColumnGrid>
              <CloseOutlinedButton type="button" id="closeButton" onClick={onClose}>
                Cancel
              </CloseOutlinedButton>
              <CreateAgentButton type="submit" id="createAgentButton" loading={loading}>
                Create agent
              </CreateAgentButton>
            </TwoColumnGrid>
          </form>
        )}
      </Formik>
    );
  }
}

CreateAgent.propTypes = propTypes;

export default CreateAgent;
