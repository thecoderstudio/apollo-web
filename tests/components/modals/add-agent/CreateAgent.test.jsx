import React from 'react';
import configureStore from 'redux-mock-store';
import waitForExpect from 'wait-for-expect';
import axios from 'axios';
import { StatusCodes } from 'http-status-codes';
import renderer from 'react-test-renderer';
import CreateAgent from '../../../src/components/modals/agg-agent/CreateAgent';
import NewAgentHandler from '../../../../src/lib/NewAgentHandler';

const mockStore = configureStore([]);
jest.mock('axios');

function getComponent(store, spy=jest.fn()) {
  return renderer.create(
    <CreateAgent onClose={spy}/>
  );
}

function submitForm(root, name) {
  const nameInput = root.findByProps({ placeholder: '007' });
  const form = root.findByType('form');

  nameInput.props.onChange({ currentTarget : {
    name: 'name',
    value: name
  }});

  form.props.onSubmit();
}

describe('create Agent', () => {
  let spy;

  it('renders correctly', () => {
    const tree = getComponent(store).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('closes modal correctly', async () => {
    const component = getComponent(store, spy);
    const instance = component.root;
    instance.findByProps({ id: 'directlyButton' }).props.onClick();
    component.toJSON();
    instance.findByProps({ id: 'closeButton' }).props.onClick();
    expect(component.toJSON()).toMatchSnapshot();

    await waitForExpect(() => {
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  it('validates input', async () => {
    const component = getComponent(store);
    const instance = component.root;
    instance.findByProps({ id: 'directlyButton' }).props.onClick();
    component.toJSON();

    submitForm(instance, '');
    submitForm(instance, new Array(102).join( 'x' ));

    await waitForExpect(() => {
      expect(axios.post).not.toHaveBeenCalled();
    });
  });

  it('creates agent', async () => {
    axios.post.mockResolvedValue({
      status: 200,
      data: { id: 'test', oauth_client: { secret: 'test' } }
    });

    const component = getFinalPageComponent(store, 'directlyButton');
    await waitForExpect(() => {
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  it('handle unsuccesful post', async () => {
    axios.post.mockImplementationOnce(() => Promise.reject({
      response: {
        status: StatusCodes.BAD_REQUEST,
        statusText: "Bad request",
        data: {}
      }
    }));

    getFinalPageComponent(store, 'directlyButton');

    axios.post.mockImplementationOnce(() => Promise.reject({
      response: {
        status: StatusCodes.INTERNAL_SERVER_ERROR,
        statusText: "Something went wrong",
        data: {}
      }
    }));

    getFinalPageComponent(store, 'directlyButton');

    await waitForExpect(() => {
      expect(axios.post).toHaveBeenCalledTimes(2);
    });
  });

  it('calls download file correctly', async () => {
    axios.post.mockResolvedValue({
      status: 200,
      data: { id: 'test', oauth_client: { secret: 'test' } }
    });

    const spy = jest.spyOn(NewAgentHandler.prototype, 'downloadFile').mockImplementation(() => { return; });
    const component = getFinalPageComponent(store, 'manualButton');
    const instance = component.root;
    component.toJSON();

    axios.get.mockResolvedValue({
      status: 200,
      data: ''
    });

    await waitForExpect(() => {
      instance.findByProps({ id: 'downloadBinaryButton' }).props.onClick();
      expect(spy).toHaveBeenCalled();
    });
  });
});