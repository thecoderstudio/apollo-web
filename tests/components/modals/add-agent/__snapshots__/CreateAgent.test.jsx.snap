// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create Agent closes modal correctly 1`] = `
.c6 {
  position: relative;
}

.c8 {
  border: none;
  border-radius: 5px;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 600;
  cursor: pointer;
  padding: 15px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c9 {
  border-color: white transparent;
  border-style: solid;
  float: right;
  margin-top: 5px;
  border-width: 7px 7px 0px 7px;
  height: 0px;
  width: 0px;
  margin-left: 10px;
  -webkit-transform: rotate(0);
  -ms-transform: rotate(0);
  transform: rotate(0);
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
}

.c12 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
  width: 100%;
  height: 50px;
  margin: auto;
  display: block;
}

.c12:hover {
  opacity: 0.90;
}

.c12:focus {
  outline: none;
}

.c12:active {
  opacity: 0.80;
}

.c12:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c3 {
  height: 80px;
}

.c5 {
  min-height: 50px;
  width: 100%;
  border: none;
  border-radius: 5px;
  font-family: 'Libre Franklin',sans-serif;
  font-size: 1rem;
  padding-left: 10px;
  opacity: 0.9;
  border: 1px solid transparent;
}

.c5:focus {
  outline: none;
  border: 1px solid;
}

.c10 {
  display: grid;
  grid-template-columns: 50% 50%;
  margin: 20px 0px 20px 0px;
}

.c0 {
  display: grid;
  width: 100%;
  grid-template-columns: [text] 50% [dropdown] 50%;
  height: 100px;
}

.c2 {
  grid-column: dropdown;
}

.c7 {
  width: 200px;
  float: right;
}

.c1 {
  grid-column: text;
  line-height: 75px;
  margin-top: 0px;
  margin-bottom: 0px;
}

.c13 {
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
}

.c4 {
  font-family: 'B612',sans-serif;
  font-weight: 600;
  width: 200px;
  float: right;
}

.c11 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
  background-color: transparent;
  border: 1px solid;
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
}

.c11:hover {
  opacity: 0.90;
}

.c11:focus {
  outline: none;
}

.c11:active {
  opacity: 0.80;
}

.c11:disabled:hover {
  opacity: 1;
  cursor: auto;
}

@media (max-width:47.9375em) {
  .c10 {
    grid-template-columns: [column-one] 100%;
    grid-template-rows: [row-one] 1fr [row-two] 1fr;
  }
}

@media (max-width:47.9375em) {
  .c0 {
    height: 120px;
    grid-template-columns: [column-one] 100%;
    grid-template-rows: [text] 30px [dropdown] 85px;
  }
}

@media (max-width:47.9375em) {
  .c2 {
    grid-column: column-one;
    grid-row: dropdown;
    width: 100%;
  }
}

@media (max-width:47.9375em) {
  .c7 {
    width: 100%;
  }
}

@media (max-width:47.9375em) {
  .c1 {
    width: 100%;
    line-height: 100%;
    grid-column: column-one;
    grid-row: text;
  }
}

@media (max-width:47.9375em) {
  .c13 {
    width: 100%;
    margin-top: 15px;
  }
}

@media (max-width:47.9375em) {
  .c4 {
    width: 100%;
    grid-row: row-two;
  }
}

@media (max-width:47.9375em) {
  .c11 {
    width: 100%;
    grid-row: row-two;
  }
}

<form
  onSubmit={[Function]}
>
  <div
    className="c0"
  >
    <p
      className="c1"
    >
      Agent name
    </p>
    <div
      className="c2"
    >
      <div
        className="c3 c4"
      >
        <input
          className="c5"
          name="name"
          onChange={[Function]}
          placeholder="007"
          type="name"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="c0"
  >
    <p
      className="c1"
    >
      Operating system
    </p>
    <div
      className="c2"
    >
      <div
        className="c6 c7"
      >
        <div
          className="c8"
          id="dropdown"
          onClick={[Function]}
        >
          linux
          <div
            className="c9"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="c0"
  >
    <p
      className="c1"
    >
      Architecture
    </p>
    <div
      className="c2"
    >
      <div
        className="c6 c7"
      >
        <div
          className="c8"
          id="dropdown"
          onClick={[Function]}
        >
          amd64
          <div
            className="c9"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="c10"
  >
    <button
      className="c11"
      id="closeButton"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      type="button"
    >
      Cancel
    </button>
    <button
      className="c12 c13"
      onClick={null}
    >
      Create agent
    </button>
  </div>
</form>
`;

exports[`create Agent renders correctly 1`] = `
.c6 {
  position: relative;
}

.c8 {
  border: none;
  border-radius: 5px;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 600;
  cursor: pointer;
  padding: 15px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c9 {
  border-color: white transparent;
  border-style: solid;
  float: right;
  margin-top: 5px;
  border-width: 7px 7px 0px 7px;
  height: 0px;
  width: 0px;
  margin-left: 10px;
  -webkit-transform: rotate(0);
  -ms-transform: rotate(0);
  transform: rotate(0);
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
}

.c12 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
  width: 100%;
  height: 50px;
  margin: auto;
  display: block;
}

.c12:hover {
  opacity: 0.90;
}

.c12:focus {
  outline: none;
}

.c12:active {
  opacity: 0.80;
}

.c12:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c3 {
  height: 80px;
}

.c5 {
  min-height: 50px;
  width: 100%;
  border: none;
  border-radius: 5px;
  font-family: 'Libre Franklin',sans-serif;
  font-size: 1rem;
  padding-left: 10px;
  opacity: 0.9;
  border: 1px solid transparent;
}

.c5:focus {
  outline: none;
  border: 1px solid;
}

.c10 {
  display: grid;
  grid-template-columns: 50% 50%;
  margin: 20px 0px 20px 0px;
}

.c0 {
  display: grid;
  width: 100%;
  grid-template-columns: [text] 50% [dropdown] 50%;
  height: 100px;
}

.c2 {
  grid-column: dropdown;
}

.c7 {
  width: 200px;
  float: right;
}

.c1 {
  grid-column: text;
  line-height: 75px;
  margin-top: 0px;
  margin-bottom: 0px;
}

.c13 {
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
}

.c4 {
  font-family: 'B612',sans-serif;
  font-weight: 600;
  width: 200px;
  float: right;
}

.c11 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
  background-color: transparent;
  border: 1px solid;
  width: 200px;
  height: 50px;
  margin: auto;
  display: block;
}

.c11:hover {
  opacity: 0.90;
}

.c11:focus {
  outline: none;
}

.c11:active {
  opacity: 0.80;
}

.c11:disabled:hover {
  opacity: 1;
  cursor: auto;
}

@media (max-width:47.9375em) {
  .c10 {
    grid-template-columns: [column-one] 100%;
    grid-template-rows: [row-one] 1fr [row-two] 1fr;
  }
}

@media (max-width:47.9375em) {
  .c0 {
    height: 120px;
    grid-template-columns: [column-one] 100%;
    grid-template-rows: [text] 30px [dropdown] 85px;
  }
}

@media (max-width:47.9375em) {
  .c2 {
    grid-column: column-one;
    grid-row: dropdown;
    width: 100%;
  }
}

@media (max-width:47.9375em) {
  .c7 {
    width: 100%;
  }
}

@media (max-width:47.9375em) {
  .c1 {
    width: 100%;
    line-height: 100%;
    grid-column: column-one;
    grid-row: text;
  }
}

@media (max-width:47.9375em) {
  .c13 {
    width: 100%;
    margin-top: 15px;
  }
}

@media (max-width:47.9375em) {
  .c4 {
    width: 100%;
    grid-row: row-two;
  }
}

@media (max-width:47.9375em) {
  .c11 {
    width: 100%;
    grid-row: row-two;
  }
}

<form
  onSubmit={[Function]}
>
  <div
    className="c0"
  >
    <p
      className="c1"
    >
      Agent name
    </p>
    <div
      className="c2"
    >
      <div
        className="c3 c4"
      >
        <input
          className="c5"
          name="name"
          onChange={[Function]}
          placeholder="007"
          type="name"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="c0"
  >
    <p
      className="c1"
    >
      Operating system
    </p>
    <div
      className="c2"
    >
      <div
        className="c6 c7"
      >
        <div
          className="c8"
          id="dropdown"
          onClick={[Function]}
        >
          linux
          <div
            className="c9"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="c0"
  >
    <p
      className="c1"
    >
      Architecture
    </p>
    <div
      className="c2"
    >
      <div
        className="c6 c7"
      >
        <div
          className="c8"
          id="dropdown"
          onClick={[Function]}
        >
          amd64
          <div
            className="c9"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="c10"
  >
    <button
      className="c11"
      id="closeButton"
      onClick={[MockFunction]}
      type="button"
    >
      Cancel
    </button>
    <button
      className="c12 c13"
      onClick={null}
    >
      Create agent
    </button>
  </div>
</form>
`;
